servicetest/testframework/lib.py

```diff
@@ -32,9 +32,110 @@
 import Queue
 import subprocess
 import dbus.mainloop.glib
+import os.path
 from gi.repository import GObject
 
 
+class ConfigFile(object):
+    """ Represents the configuration file for SoftwareContainer
+        All keys and values have to be strings.
+
+        The format of a config file is as follows:
+
+        [Group1]
+        item1 = value1
+        item2 = value2
+
+        [Group2]
+        item1 = value1
+        item2 = value2
+
+        Initialization is done as follows:
+
+        config = ConfigFile("/where/to/save/me", {
+            "Group1": {
+                "item1": "value1",
+                "item2": "value2"
+            },
+
+            "Group2": {
+                "item1": "value1",
+                "item2": "value2"
+            }
+        })
+    """
+
+    def __init__(self, location, full_config):
+        """ Create a config object from a given dict of dicts.
+        """
+        self.__path = location
+        self.__groups = full_config
+
+    def path(self):
+        return self.__path
+
+    def config_as_string(self):
+        """ Convert this object into a string that can be written to a config file
+        """
+
+        # Copy the values from the defaults to the given config
+        # Only add them if there is no matching value already in the given config.
+        defaults = self.__default_config()
+        for group in defaults:
+            if group not in self.__groups:
+                self.__groups[group] = defaults[group]
+            else:
+                for item in defaults[group]:
+                    if item not in self.__groups[group]:
+                        self.__groups[group][item] = defaults[group][item]
+
+        result = ""
+        for group in self.__groups:
+            result += "[" + group + "]\n"
+            for item in self.__groups[group]:
+                result += item + " = " + self.__groups[group][item] + "\n"
+
+        return result
+
+    def __default_config(self):
+        """ Read the softwarecontainer-config prepared by cmake, and return its values as a dict.
+            This expects the file to be available in the same directory as this file.
+
+            The file, and the dict, will only contain values that are mandatory options, and as
+            such has no default values within SoftwareContainer. These are the values that one
+            has to have in a config file for the agent to start at all. Any other config values
+            have reasonable defaults and are therefore commented out in the config file.
+        """
+        config_path = os.path.join(os.path.dirname(__file__), "softwarecontainer-config")
+
+        # Open the file, read it and close it
+        fh = open(config_path, "r")
```

Comment:
Nitpicking here so do as you please, but it's more common in our code to use "with". I guess you want to avoid it because it looks weird to use the variable file_contents later outside of the "with"? It's perfectly OK afaik in Python since there is no new scope in a "with".